packages/flutter/lib/src/material/search.dart

```diff
@@ -0,0 +1,416 @@
+  String get query => _textEditingController.text;
+  set query(String value) {
+    assert(query != null);
+    _textEditingController.text = value;
+  }
+
+  /// Transition to show search results.
+  ///
+  /// If the user taps on a suggestion provided by [buildSuggestions] the
+  /// screen should typically transition to the page showing the search
+  /// results for the suggested query. This transition can be triggered
+  /// by calling this method.
+  ///
+  /// See also:
+  ///
+  ///  * [showSuggestions] to show the search suggestions again.
+  ///  * [currentBody] to learn whether the results or the suggestions are
+  ///    currently shown.
+  void showResults(BuildContext context) {
+    focusNode.unfocus();
+    _currentBody.value = SearchBody.results;
+  }
+
+  /// Transition to show the search suggestions and places the input focus
+  /// back into the search text field.
+  ///
+  /// If the results are currently shown this method can be used to go back
+  /// to showing the search suggestions.
+  ///
+  /// See also:
+  ///
+  ///  * [showResults] to show the search results.
+  ///  * [currentBody] to learn whether the results or the suggestions are
+  ///    currently shown.
+  void showSuggestions(BuildContext context) {
+    FocusScope.of(context).requestFocus(focusNode);
+    _currentBody.value = SearchBody.suggestions;
+  }
+
+  /// The body that is currently shown under the [AppBar] in the search page.
+  ///
+  /// Will be null if the search page is currently not shown.
+  ///
+  /// See also:
+  ///
+  ///  * [showResults] to show the search results.
+  ///  * [showSuggestions] to show the search suggestions.
+  SearchBody get currentBody => _currentBody.value;
+
+  /// Closes the search page and returns to the underlying route.
+  ///
+  /// The value provided for `result` is used as the return value of the call
+  /// to [showSearch] that launched the search initially.
+  void close(BuildContext context, T result) {
+    _currentBody.value = null;
+    focusNode.unfocus();
+    _result.complete(result);
+    Navigator.of(context).pop(result);
+  }
+
+  /// [Animation] triggered while the search pages fades in or out.
+  ///
+  /// This animation is commonly used to animate [AnimatedIcon]s of
+  /// [IconButton]s return by [buildLeading] or contained within the route
+  /// below the search page.
+  Animation<double> get transitionAnimation => _proxyAnimation;
+
+  /// [FocusNode] used by the text field showing the current search query.
+  final FocusNode focusNode = new FocusNode();
+
+  final TextEditingController _textEditingController = new TextEditingController();
+
+  final ProxyAnimation _proxyAnimation = new ProxyAnimation(kAlwaysDismissedAnimation);
+
+  final ValueNotifier<SearchBody> _currentBody = new ValueNotifier<SearchBody>(null);
+
+  Completer<T> _result;
+
+}
+
+/// Describes the body that is currently shown under the [AppBar] in the
+/// search page.
+enum SearchBody {
```

Comment:
Same as earlier: do SearchDelegate subclassers need this?

Reply:
Same.